File: test/cfi_allocate/cfi_allocate.c
// https://stackoverflow.com/a/37837472
#include <ISO_Fortran_binding.h>

void alloc3d(CFI_cdesc_t* x) {
  if (x->base_addr) CFI_deallocate(x);

  CFI_index_t lower[3]={-1, 5, 1};
  CFI_index_t upper[3]={20, 9, 5};

  CFI_allocate(x, lower, upper, 0);
}

void dealloc3d(CFI_cdesc_t* x) {
  CFI_deallocate(x);
}

File: test/iso_fortran_binding/cfi_allocate.f90
!! https://stackoverflow.com/a/37837472
program alloc3

use, intrinsic :: iso_c_binding, only : c_float

implicit none

interface
subroutine alloc3d(x) bind(c)
  import c_float
  real(c_float), allocatable, intent(inout) :: x(:,:,:)
end subroutine

subroutine dealloc3d(x) bind(c)
  import c_float
  real(c_float), allocatable, intent(inout) :: x(:,:,:)
end subroutine
end interface

integer :: i
real(c_float), allocatable :: x(:,:,:)

allocate(x(6,2,5))
print *,"x has been allocated like x(", (LBOUND(x,i), UBOUND(x,i), i=1,3), ")"
!! C code will deallocate x

call alloc3d(x)
if (allocated(x)) then
  print *, "x has been allocated like x(", (LBOUND(x,i), UBOUND(x,i), i=1,3), ")"
endif

call dealloc3d(x)
if (allocated(x)) error stop "CFI_deallocate failed"

print *, "OK: CFI_allocate demo"

end program
